chainer/variable.py

```diff
@@ -50,6 +51,134 @@ def make_message(message):
         raise ValueError(make_message(msg))
 
 
+class _VariableData(object):
+
+    initializer = None
+    _grad_initializer = None
+    _initial_device = -1
+
+    def __init__(self, data=None, grad=None, initializer=None):
+        if data is None:
+            self._initializer = (
+                initializers.NaN() if initializer is None else initializer)
+            dtype = getattr(self.initializer, 'dtype', numpy.float32)
+            self._grad_initializer = initializers.NaN(dtype)
+        elif not isinstance(data, (numpy.ndarray, cuda.ndarray)):
+                msg = '''numpy.ndarray or cuda.ndarray are expected.
+Actual: {0}'''.format(type(data))
+                raise TypeError(msg)
+
+        self.data = data
```

Comment:
How about changing `data` to `_data` as we do not expect users to access this attribute directly.